File: sdk/mobile/android/src/main/kotlin/com/summit/odps/android/TelemetryPipeline.kt
package com.summit.odps.android

/**
 * Coordinates minimisation, consent gates, batching, DP and uploading.
 */
class TelemetryPipeline(
  private val policyPack: PolicyPack,
  private val consentManager: ConsentManager,
  private val networkStateProvider: NetworkStateProvider,
  private val uploader: TelemetryUploader,
  private val batchStorage: BatchStorage = BatchStorage(),
  private val privacyEngine: DifferentialPrivacyEngine = DifferentialPrivacyEngine(),
  private val verifier: TransmissionVerifier? = null
) {

  fun track(event: TelemetryEvent) {
    if (!consentManager.hasConsent()) {
      return
    }

    val filterResult = policyPack.filter(event)
    val sanitizedMetrics = privacyEngine.sanitize(event.metrics, policyPack)
    val envelope = TransmissionEnvelope(
      event = event,
      filteredPayload = filterResult.allowedPayload,
      blockedFields = filterResult.blockedFields,
      blockedSensitiveFields = filterResult.blockedSensitiveFields,
      sanitizedMetrics = sanitizedMetrics
    )
    batchStorage.enqueue(envelope)
    tryFlush()
  }

  fun flush() {
    tryFlush()
  }

  private fun tryFlush() {
    if (!networkStateProvider.isNetworkAvailable()) {
      return
    }

    val batch = batchStorage.drain()
    if (batch.isEmpty()) {
      return
    }

    val result = uploader.upload(batch)
    if (!result.success) {
      batchStorage.requeue(batch)
      return
    }

    verifier?.let { v ->
      batch.forEach { v.recordTransmission(it, policyPack) }
    }
  }
}

File: sdk/mobile/ios/Sources/ODPS/TelemetryPipeline.swift
public final class TelemetryPipeline {
    private let policyPack: PolicyPack
    private let consentManager: ConsentManager
    private let networkProvider: NetworkStateProvider
    private let uploader: TelemetryUploader
    private let storage: BatchStorage
    private let privacyEngine: DifferentialPrivacyEngine
    private let verifier: TransmissionVerifier?

    public init(policyPack: PolicyPack,
                consentManager: ConsentManager,
                networkProvider: NetworkStateProvider,
                uploader: TelemetryUploader,
                storage: BatchStorage = BatchStorage(),
                privacyEngine: DifferentialPrivacyEngine = DifferentialPrivacyEngine(),
                verifier: TransmissionVerifier? = nil) {
        self.policyPack = policyPack
        self.consentManager = consentManager
        self.networkProvider = networkProvider
        self.uploader = uploader
        self.storage = storage
        self.privacyEngine = privacyEngine
        self.verifier = verifier
    }

    public func track(event: TelemetryEvent) {
        guard consentManager.hasConsent() else { return }
        let filterResult = policyPack.filter(event: event)
        let sanitizedMetrics = privacyEngine.sanitize(metrics: event.metrics, pack: policyPack)
        let envelope = TransmissionEnvelope(event: event,
                                            filteredPayload: filterResult.allowedPayload,
                                            blockedFields: filterResult.blockedFields,
                                            blockedSensitiveFields: filterResult.blockedSensitiveFields,
                                            sanitizedMetrics: sanitizedMetrics)
        storage.enqueue(envelope)
        flush()
    }

    public func flush() {
        guard networkProvider.isNetworkAvailable() else { return }
        let batch = storage.drain()
        guard !batch.isEmpty else { return }
        let result = uploader.upload(batch: batch)
        guard result.success else {
            storage.requeue(batch)
            return
        }
        verifier?.record(batch: batch, pack: policyPack)
    }
}
